### ISSUE TITLE
Offer OSS alternatives to Redis

### ISSUE THREAD

**Issue by todaywasawesome** [argoproj/argo-cd#17970]
# Summary

The [Redis License is changing](https://redis.io/blog/redis-adopts-dual-source-available-licensing/) from BSD-3-Clause to source available BUSL which does not open source under the [OSI definition](https://opensource.org/osd). We should consider open-source alternatives as a default replacement for Redis.

# Motivation

First off, let's recognize @hairmare for bringing the suggestion to switch to ValKey in #17892. There are several reasons we should consider a replacement:

* Deploying Argo CD should only require OSS software. 
* There are potential performance improvements to be made with an alternative. 
* Some ambiguity about CNCF license compatibility. 

# Proposal
There are at least two potential drop-in replacements for Redis that would still maintain compatibility with the Redis API. 
* [Valkey](https://valkey.io/) 
* [KeyDB](https://docs.keydb.dev/)

Valkey is a CNCF fork of existing Redis BSD-3-Clause meant to be a drop in replacement that will maintain compatibility.

KeyDB is a performance-focused fork with several years of commits. It's primarily based on Redis 6 and does anticipate some impact from the [Redis License change](https://github.com/Snapchat/KeyDB/issues/798).

We should evaluate all potential options, including changing the way we leverage Redis today. We need a spike to review the implications of using these alternatives and measure performance impacts. In the meantime, we should keep in mind:

* Argo CD is several versions behind on Redis as it is and security updates are going to be offered for supported BSD versions of Redis until Redis Community Edition 9.0 which should arrive sometime in 2026. We have time to make an informed decision.
* The CNCF [3rd party license policy](https://github.com/cncf/foundation/blob/main/allowed-third-party-license-policy.md) does not permit BUSL however I believe these 3rd party license issues only apply to libraries we build and package and not to binaries in a helm chart for instance.
* There are many Argo CD users using enterprise versions of Redis.

**Comment by joebowbeer** [comment]
Affected manifests:
* [manifests/base/redis](https://github.com/argoproj/argo-cd/tree/master/manifests/base/redis)
* [manifests/ha/base/redis-ha](https://github.com/argoproj/argo-cd/tree/master/manifests/ha/base/redis-ha)
  * https://github.com/DandyDeveloper/charts/tree/master/charts/redis-ha

**Comment by todaywasawesome** [comment]
Worth adding that the result may not be replacing redis at all. We're not required to by the license.

**Comment by rumstead** [comment]
> Worth adding that the result may not be replacing redis at all. We're not required to by the license.

Sharing the discussion where it was brought up

https://github.com/argoproj/argo-cd/discussions/17599#discussioncomment-8901948

**Comment by madolson** [comment]
Hey, I'm one of the maintainers of Valkey. I just wanted to drop by to add some more context. Feel free to @me with questions if you have any.

> Argo CD is several versions behind on Redis as it is and security updates are going to be offered for supported BSD versions of Redis until Redis Community Edition 9.0 which should arrive sometime in 2026. 

I want to highlight that most of the top contributors are no longer working on Redis Community Edition, and moved to Valkey. Yes there will be support, but we already have fixes in our 7.2.5 that weren't present in their latest 7.2 release. Also it seems you are on 7.0, which will have its support dropped sooner (Unless they changed their support policy, it'll be in 2025).

> We have time to make an informed decision.

I think this is the right framing, don't rush the decision. My guidance so far has been to wait until you would normally do an upgrade and then treat the move to Valkey as a minor version bump. We are fully compatible, so it's easy to test on, but take your time.

**Comment by ramanNarasimhan77** [comment]
Hello,
I found following matches for "redis" in Argo-CD's go.mod from the master branch:

github.com/alicebob/miniredis/v2 v2.33.0
github.com/go-redis/cache/v9 v9.0.0
github.com/redis/go-redis/v9 v9.6.1

I observed that Valkey has a separate go client - https://github.com/valkey-io/valkey-go?tab=readme-ov-file#features 
Will the move to Valkey also require changes in Argo-CD to use a different client?

**Comment by madolson** [comment]
As of today, the same go client that works with Redis will also work with Valkey. Both Redis and Valkey use the same RESP protocol, so as long as you aren't using any of the new features it should be a drop in replacement. I quickly checked the packages, and they should all work except miniredis. Miniredis is a "fake" redis used for testing, so it should be fine to use.

**Comment by james-callahan** [comment]
Is there any progress on this? The old version of redis currently used by argocd's kustomizations is starting to get flagged by update tools as significantly out of date.

**Comment by venkatamutyala** [comment]
ValKey is 8.0 is GA as of last month and a patch has been dropped in the past week (8.0.1). It should be fully compatible with redis 7.X.X.

I think switching to a compatible alternative like ValKey sooner will help keep it a low risk change versus doing it later.

**Comment by venkatamutyala** [comment]
@todaywasawesome is this something a corporate sponsor like Akuity or Codefresh would handle on it's own time/schedule, or is there anything we as OSS/individual contributors can do to help push this forward?

**Comment by crenshaw-dev** [comment]
afaik there's no reason we can't start making changes upstream. I know the redis-ha setup is complex, and I'm not sure if we can take advantage of that with an alternative tool or if we will have to rebuild or find an alternative for it.

**Comment by todaywasawesome** [comment]
> @todaywasawesome is this something a corporate sponsor like Akuity or Codefresh would handle on it's own time/schedule, or is there anything we as OSS/individual contributors can do to help push this forward? 

No, I think anyone can contribute. First, I'd like to give an option for people to opt into Valley or even just docs on how to replace. 

Then I think adding some tests to start offering support makes sense. Afaik, people can keep using Redis with its license change and I know we have users using the enterprise version with Argo CD.

**Comment by reegnz** [comment]
My team has opted into replacing Redis 7.2.x with Valkey 7.2.x (although we're not running in HA). So far it is definitely a drop-in replacement, haven't found any issues with it.
Migration was literally just switching the redis docker image to `valkey/valkey:7.2.7`, not sure if there's any other migration doc to be written, it's a recent redis fork.

I assume the HA setup would be a similar experience, but someone else has to verify that. 

I'd not jump onto the 8.x train immediatelly, as right now 7.2 branch is still getting security patches.

The ArgoCD project should make sure that everything is migrated to Valkey 7.2.7, as that's the closest to the redis version [referenced by the argocd helm chart today](https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/values.yaml#L1277). 

In the meantime work on a branch on Valkey 8.x adoption for 2.13.0?

I don't mind people using Valkey or Redis or whatever as long as it's api-compatible so replacing it is trivial. 

But the ArgoCD project as a CNCF project should default to an infrastructure component that has a CNCF-compatible icense and keeps getting security patches under a CNCF-compatible license. Right now that is not met with the new license of redis.

**Comment by vitaliyf** [comment]
Sadly, redis got automatically updated to 7.4.1 in https://github.com/argoproj/argo-helm/pull/3023 which not only moved it to the new https://github.com/redis/redis/blob/7.4.0/LICENSE.txt but also made it non-upgradable to Valkye (see https://github.com/valkey-io/valkey/issues/845)

If you are still running with Redis 7.2 then the switch to Valkye 7.2 (or 8.0) is a trivial drop-in replacement.

**Comment by todaywasawesome** [comment]
> Sadly, redis got automatically updated to 7.4.1 in https://github.com/argoproj/argo-helm/pull/3023 which not only moved it to the new https://github.com/redis/redis/blob/7.4.0/LICENSE.txt but also made it non-upgradable to Valkye (see https://github.com/valkey-io/valkey/issues/845)
> 
> If you are still running with Redis 7.2 then the switch to Valkye 7.2 (or 8.0) is a trivial drop-in replacement.

Since the Redis cache is throw away in Argo CD, it shouldn't be an issue to just rebuild the cache with Valkey. Did you try switching the image after 7.4.1 and did it throw an error?

**Comment by vitaliyf** [comment]
If you are still running on Redis 7.2
- Simply set redis-ha image to valkey (i.e., `--set redis-ha.image.repository docker.io/valkey/valkey --set redis-ha.image.tag 7.2-alpine`)

If you already upgraded to Redis 7.4
- If you attempt to do the above you will get errors mentioned in https://github.com/valkey-io/valkey/issues/845 (`Can't handle RDB format version 12` and `Primary does not understand REPLCONF VERSION: -ERR Unrecognized REPLCONF option: version`)
- Since the cache is throw-away the fix is to discard existing Redis data. The simplest method to do that for me was to start ArgoCD without redis-ha (`--set redis-ha.enabled false`), wait for a moment as it terminates redis-ha pods, then re-deploy with redis-ha enabled and repository set to valkey.

PS. We also verified that upgrading to valkey 8.0 is safe and easy (`--set redis-ha.image.tag 8.0.1-alpine`)

**Comment by todaywasawesome** [comment]
@vitaliyf that would make a great bit of documentation if you're keen to contribute it.

**Comment by todaywasawesome** [comment]
@vitaliyf Looking again at the manifests, there's no persistence. When you upgrade Redis, I think it literally dumps everything and recreates it. So I don't see how it would create an issue to switch between Redis and Valkey given that there shouldn't be a database upgrade at all. 

You're verifying you tried it with Argo CD and it caused errors? Can you share the logs?

**Comment by sathieu** [comment]
What is the status of this ticket?

Has the alternative be choosen? valkey or keydb?

Should we go ahead and add a PR to switch?

**Comment by madolson** [comment]
Just FYI, the main maintainer of KeyDB left snap recently. https://github.com/Snapchat/KeyDB/issues/895

**Comment by andrii-korotkov-verkada** [comment]
I've filed https://github.com/argoproj/argo-cd/issues/22108 which is probably good to do anyways regardless of a chosen path.

**Comment by reegnz** [comment]
Momentum is gaining with valkey adoption and some distros already started completely dropping redis:
* https://fedoraproject.org/wiki/Changes/Replace_Redis_With_Valkey#Replace_Redis_with_Valkey
* https://archlinux.org/news/valkey-to-replace-redis-in-the-extra-repository/

According to https://endoflife.date/redis the EOL for security updates is 28 Feb 2026.
Would be nice to not do this last minute.
Switching to 7.4 and higher is not an option for most users IMHO, as it's not FLOSS anymore.

**Comment by DimitarKapashikov** [comment]
Hi All, what will change with the recent Redis license change - https://thenewstack.io/redis-is-open-source-again/ ? Can we still use Redis in ArgoCD?

**Comment by reegnz** [comment]
I think it's unwise to use a project where there is a persistent risk of a license change whenever the governing entity (in this case redis inc) switches CEO-s or goes public etc.

The trust in that regard has completely eroded and the current switch to AGPL doesn't bring back that trust.

Given that ArgoCD is a CNCF project I'd prefer it use a project that either has a CNCF compatible license, or a license exception for a project that is actually foundation backed. AFAICT AGPL is not an approved license so there must at least be an exception by the CNCF.

With Valley being backed by the Linux Foundation I don't understand what the hesitation is in adopting it to replace redis.

**Comment by reegnz** [comment]
I think it's unwise to use a project where there is a persistent risk of a license change whenever the governing entity (in this case redis inc) switches CEO-s or goes public etc.

The trust in that regard has completely eroded and the current switch to AGPL doesn't bring back that trust.

Given that ArgoCD is a CNCF project I'd prefer it use a project that either has a CNCF compatible license, or a license exception for a project that is actually foundation backed. AFAICT AGPL is not an approved license so there must at least be an exception by the CNCF.

With Valley being backed by the Linux Foundation I don't understand what the hesitation is in adopting it to replace redis.

EDIT: the CNCF policy on licenses of software dependencies: https://github.com/cncf/foundation/blob/main/allowed-third-party-license-policy.md
I'd assume this also applies to service dependencies on docker images referenced in helm charts.

**Comment by todaywasawesome** [comment]
@reegnz @mkilchhofer @DimitarKapashikov There has been a lot of confusion about the Redis license change and it's impact on Argo CD. I saw this PR https://github.com/argoproj/argo-helm/pull/3271 to Argo Helm downgrading Helm to update the license. In this case, Helm was actually ahead of upstream Argo CD which is still on 7.2. I think we should have a policy that the Helm Chart follow upstream versions as it creates release confusion when they don't. 

**The CNCF License Policy on Software Dependencies does not apply to packaged helm charts.**
From the policy (emphasis added):
> A third-party component under a non-Apache 2.0 license is deemed automatically approved by the Governing Board for **inclusion in a CNCF codebase** as an exception to the CNCF Intellectual Property Policy, if all of the following apply:

This policy applies to components that we use in our codebase, things like libraries etc. Those things have an impact on our license but a reference in a Kubernetes manifest does not constitute a component in the codebase. Today, a user can replace the Redis version with their own version, including Redis Enterprise (which is common) or Valkey (which is becoming more common). They can also deploy without Redis at all, though Argo CD will be less performant without access to the cache. These are user choices. 

My suggestion would be we simply document how to replace Redis with Valkey or another option, and potentially include a flag to make that easy. I'll bring this up for discussion in this week's contributor's meeting. Redis is now moving to AGPL v3, because we do not modify or interact with the Redis code in anyway, I do not believe it has any impact on us.

**Comment by rajivml** [comment]
>Redis is now moving to AGPL v3, because we do not modify or interact with the Redis code in anyway, I do not believe it has any impact on us.
AGPL v3 is also bad where one cannot redistribute argoCD if argoCD is bundled along with their software

**Comment by rajivml** [comment]
hi, is there any latest decision on if Redis is being replaced by valkey or not ?

**Comment by trashhalo** [comment]
> [@reegnz](https://github.com/reegnz) [@mkilchhofer](https://github.com/mkilchhofer) [@DimitarKapashikov](https://github.com/DimitarKapashikov) There has been a lot of confusion about the Redis license change and it's impact on Argo CD. I saw this PR [argoproj/argo-helm#3271](https://github.com/argoproj/argo-helm/pull/3271) to Argo Helm downgrading Helm to update the license. In this case, Helm was actually ahead of upstream Argo CD which is still on 7.2. I think we should have a policy that the Helm Chart follow upstream versions as it creates release confusion when they don't.
> 
> **The CNCF License Policy on Software Dependencies does not apply to packaged helm charts.** From the policy (emphasis added):
> 
> > A third-party component under a non-Apache 2.0 license is deemed automatically approved by the Governing Board for **inclusion in a CNCF codebase** as an exception to the CNCF Intellectual Property Policy, if all of the following apply:
> 
> This policy applies to components that we use in our codebase, things like libraries etc. Those things have an impact on our license but a reference in a Kubernetes manifest does not constitute a component in the codebase. Today, a user can replace the Redis version with their own version, including Redis Enterprise (which is common) or Valkey (which is becoming more common). They can also deploy without Redis at all, though Argo CD will be less performant without access to the cache. These are user choices.
> 
> My suggestion would be we simply document how to replace Redis with Valkey or another option, and potentially include a flag to make that easy. I'll bring this up for discussion in this week's contributor's meeting. Redis is now moving to AGPL v3, because we do not modify or interact with the Redis code in anyway, I do not believe it has any impact on us.

I get the licensing perspective but I think we're missing a practical issue here. Staying on Redis 7.2 means we're shipping with gosu in the container, which brings 70+ CVEs (3 critical, 31 high, 27 medium). Yeah the actual exploitability of gosu vulns is [debated](https://github.com/tianon/gosu/blob/master/SECURITY.md) but that doesn't matter when you're trying to get through compliance scans.

Both [Redis 8](https://github.com/redis/docker-library-redis/pull/435) and [Valkey 7.2+](https://github.com/valkey-io/valkey-container/issues/22) have removed gosu from their dockerfiles which cleans up all those alerts. When every org deploying argocd with the helm chart has to immediately swap out the redis image to pass security scans, that's creating unnecessary friction for everyone.

The license concern is valid but the security posture shouldn't be ignored either. Even if you document how to swap to valkey/redis8, having a secure default would be better than making everyone do extra work to not fail compliance checks.

TLDR; status quo creates busy work for a lot of teams. Please upgrade default to redis 8 or valkey 7.2.x.

**Comment by lme-nca** [comment]
Hi, any update ?  I think I speak for everyone on this if I say ANY DECISION is better than NO DECISION!

**Comment by dudinea** [comment]
There was a discussion on this topic at the Contributor Experience Meeting 2025-07-10,
see here: https://www.youtube.com/watch?v=YU83mQzygYE from 2:56. 

The decisions were essencially:
 
* The new Redis license does not directly affect ArgoCD because it does not compete with Redis, even for companies that would offer Enterprise Argo.
* "We're good to go to Redis 8" 
* "and we can have a separate effort to offer support Valkey"

**Comment by icelynjennings** [comment]
Nothing to add, but for anyone curious about running `valkey` + `argocd` in a production setting, I have been doing this at my employer's org for over a month now without issue.

**Comment by james-callahan** [comment]
It seems that Argo CD 3.2 included an update to redis 8.2 :( https://github.com/argoproj/argo-cd/pull/24523

**Comment by Skaronator** [comment]
I suggest we update the scope of this issue.

First, KeyDB appears to be unmaintained. The last commit was over 18 months ago, and the last release was 2 years ago. I recommend we remove it from consideration.

Second, I propose we change the issue title to be more direct: "Proposal: Adopt Valkey as the default datastore."

There is strong strategic alignment. Valkey is a Linux Foundation (LF) project, and Argo CD is a Cloud Native Computing Foundation (CNCF) project (which is part of the LF). Adopting Valkey ensures Argo CD's core datastore dependency is governed by the same open, community-driven principles.

I recommend we replace Redis with Valkey as the default datastore in both the default installation and the official Helm chart. Redis can be maintained as a supported, non-default option, but Valkey should be the first-party, recommended choice.

**Comment by todaywasawesome** [comment]
@Skaronator Before we make Valkey the default choice, we should just add it as a choice. People can swap them out today, but there's not documentation on how (it's just [overriding the image](https://github.com/codefresh-io/argo-helm/blob/e094a1d2bbb1788a5bf98468648996db193661eb/charts/argo-cd/values.yaml#L1502)), and then adding a flag to opt into valkey, and eventually, if enough people adopt it then we can switch the default.

I don't think we have any need to put our thumb on the scale here to say people should use Valkey over what they've been using. Let them adopt it, if they do, we'll change the default.

**Comment by joebowbeer** [comment]
In addition to the [community-maintained helm chart](https://argo-cd.readthedocs.io/en/stable/operator-manual/installation/#helm), several manifests in this repo are affected:

https://github.com/argoproj/argo-cd/issues/17970#issuecomment-2078067943

redis-ha is currently vendored in this repo.

https://github.com/DandyDeveloper/charts/issues/297

**Comment by sousa-miguel** [comment]
Is there any update? I think it would be nice to offer the user the chance to opt into `valkey` before making it the default store. 
We have been using it on production for 3 months without any issues.

**Comment by venkatamutyala** [comment]
I'm happen to take it on and make a PR to the argocd helm chart. Should i open an issue there or do we want to update libraries too or just mix and match redis client libraries to valkey backend?

**Comment by joebowbeer** [comment]
@venkatamutyala If the enhancement can be made entirely in [argo-helm](https://github.com/argoproj/argo-helm) then I'd open an issue there.

**Comment by venkatamutyala** [comment]
@sousa-miguel could you confirm? Any changes required in this repo? If not, i can start an issue in https://github.com/argoproj/argo-helm

**Comment by Crayeth** [comment]
Hi, we're replacing our redises with valkey would be great if the chart supports this choice any updates?
